pkg/controller.go

```diff
@@ -127,46 +144,101 @@ func (c *controller) Run(ctx context.Context) error {
+	for pod := range c.pauseChan {
+		err := c.evictPod(ctx, pod)
 		if err != nil {
 			c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
 			klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
 			continue
 		}
 
 		if c.opts.EvictionPause > 0 {
+			klog.V(2).Infof("Pausing pod eviction for %s", c.opts.EvictionPause)
 			time.Sleep(c.opts.EvictionPause)
 		}
+
+		klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
 	}
-	return nil
 }
 
-func (c *controller) listPodsCloseToMemoryLimit(ctx context.Context) ([]*corev1.Pod, error) {
+// evict pods having a PodDisruptionBudget. If the PDB is exceeded, the eviction
+// is retried after a delay using a different channel.
+func (c *controller) evictWithPDBChanLoop(ctx context.Context) {
+	backoffChan := make(chan *corev1.Pod, c.opts.ChannelBufferSize)
+	defer close(backoffChan)
+	go func() {
+		for pod := range backoffChan {
+			attempt := 1
+		retry:
+			klog.V(2).Infof("Retrying eviction of '%s/%s' (attempt %d)", pod.Namespace, pod.Name, attempt)
+			err := c.evictPod(ctx, pod)
+			if err != nil {
+				if strings.Contains(err.Error(), "disruption budget") && attempt <= 3 {
+					time.Sleep(time.Duration(math.Pow(5, float64(attempt))) * time.Second)
+					attempt++
+					goto retry
+				}
+				c.recorder.Event(pod.DeepCopyObject(), "Warning", "SoftEviction", fmt.Sprintf("Unable to evict Pod '%s/%s' : %v", pod.Namespace, pod.Name, err))
+				klog.Errorf("error evicting '%s/%s': %v", pod.Namespace, pod.Name, err)
+				continue
+			}
+
+			klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
+		}
+	}()
+
+	for pod := range c.pdbChan {
+		err := c.evictPod(ctx, pod)
+		if err != nil {
+			backoffChan <- pod
+		}
+
+		klog.V(2).Infof("Pod '%s/%s' evicted", pod.Namespace, pod.Name)
+	}
+}
+
+func (c *controller) schedulePodEviction(ctx context.Context, pod *corev1.Pod) {
+	klog.V(1).Infof("Evicting Pod '%s/%s'", pod.Namespace, pod.Name)
+	c.recorder.Event(pod.DeepCopyObject(), "Normal", "SoftEviction", fmt.Sprintf("Pod '%s/%s' has at least one container close to its memory limit", pod.Namespace, pod.Name))
+
+	if c.hasDisruptionBudget(ctx, pod) {
+		if len(c.pdbChan) == cap(c.pdbChan) {
+			klog.Warningf("PDB channel is full, skipping eviction for Pod '%s/%s'", pod.Namespace, pod.Name)
+		} else {
+			c.pdbChan <- pod
+		}
+	} else {
+		if len(c.pauseChan) == cap(c.pauseChan) {
+			klog.Warningf("Pause channel is full, skipping eviction for Pod '%s/%s'", pod.Namespace, pod.Name)
+		} else {
+			c.pauseChan <- pod
+		}
+	}
+}
+
+func (c *controller) evictPod(ctx context.Context, pod *corev1.Pod) error {
+	evictionPolicy := policyv1.Eviction{
+		ObjectMeta: pod.ObjectMeta,
+	}
+	if c.opts.DryRun {
+		evictionPolicy.DeleteOptions = &metav1.DeleteOptions{
+			DryRun: []string{"All"},
+		}
+	}
+
+	return c.clientset.PolicyV1().Evictions(pod.Namespace).Evict(ctx, &evictionPolicy)
```

Comment:
Now that we're evicting Pods asynchronously, there is a higher chance we try to evict a Pod that is already gone right ? Do you know what kind of error is returned in such a case ? Could we silence it ? Can also be in another PR if you want to observe that live.

Reply:
Indeed this can happen. Before this PR the loop was blocking the next execution for long enough such that the next iteration had to wait until all pods had been sent an eviction (even sigterm).
I don't think this is too big of a deal. If it gets too verbose I'll issue another PR.